rfc/rfc-83/rfc-83.md

```diff
@@ -0,0 +1,233 @@
+
+```text
+    {
+      "name": "earliestInstantToRetain",
+      "type":["null", {
+        "type": "record",
+        "name": "HoodieActionInstant",
+        "fields": [
+          {
+            "name": "timestamp",
+            "type": "string"
+          },
+          {
+            "name": "action",
+            "type": "string"
+          },
+          {
+            "name": "state",
+            "type": "string"
+          }
+        ]
+      }],
+      "default" : null
+    },
+```
+
+We also need a unified interface/abstract-class to control the Plan behavior of the TableService including clustering and compaction.
+
+### Abstraction
+
+Use `PartitionBaseTableServicePlanStrategy` to control the behavior of getting partitions, filter partitions and generate table service plan etc.
+
+Since we want to control the logic of partition acquisition, partition filtering, and plan generation through different strategies, 
+in the first step, we need to use an abstraction to converge the logic of partition acquisition, partition filtering, and plan generation into the base strategy.
+
+```java
+package org.apache.hudi.table;
+
+import org.apache.hudi.common.engine.HoodieEngineContext;
+import org.apache.hudi.common.table.HoodieTableMetaClient;
+import org.apache.hudi.common.table.timeline.HoodieInstant;
+import org.apache.hudi.common.util.Option;
+import org.apache.hudi.config.HoodieWriteConfig;
+
+import java.io.IOException;
+import java.util.List;
+
+public abstract class PartitionBaseTableServicePlanStrategy<R,S> {
+
+  /**
+   * Generate table service plan based on given instant.
+   * @return
+   */
+  public abstract R generateTableServicePlan(Option<String> instant) throws IOException;
+
+  /**
+   * Generate table service plan based on given instant.
+   * @return
+   */
+  public abstract R generateTableServicePlan(List<S> operations) throws IOException;
+
+
+  /**
+   * Get partition paths to be performed for current table service.
+   * @param metaClient
+   * @return
+   */
+  public abstract List<String> getPartitionPaths(HoodieWriteConfig writeConfig, HoodieTableMetaClient metaClient, HoodieEngineContext engineContext);
+
+  /**
+   * Filter partition path for given fully paths.
+   * @param metaClient
+   * @return
+   */
+  public abstract List<String> filterPartitionPaths(HoodieWriteConfig writeConfig, List<String> partitionPaths);
+
+  /**
+   * Get incremental partitions from EarliestCommitToRetain to instantToRetain
+   * @param instantToRetain
+   * @param type
+   * @param deleteEmptyCommit
+   * @return
+   * @throws IOException
+   */
+  public List<String> getIncrementalPartitionPaths(Option<HoodieInstant> instantToRetain) {
+    throw new UnsupportedOperationException("Not support yet");
+  }
+
+  /**
+   * Returns the earliest commit to retain from instant meta
+   */
+  public Option<HoodieInstant> getEarliestCommitToRetain() {
+    throw new UnsupportedOperationException("Not support yet");
```

Comment:
The `IncrementalPartitionAwareStrategy` should be an user interface IMO, the only API we expose to user is the incremental partitions since last table service. So the logic of following should be removed:

1. generate plan (should be responsibility of the planner)
2. getEarliestCommitToRetain (should be responsibility of the planner within the plan executor)

And because the implementaion of compaction and clustering are quite different, maybe we just add two new interfaces: `IncrementalPartitionAwareCompactionStrategy` and `IncrementalPartitionAwareClusteringStrategy`

Reply:
1.  generate plan and getEarliestCommitToRetain is removed.
2. As for base abstraction, although the implementation of compaction and clustering are quite different, but for Partition Aware's Compaction and clustering, they both have the same partition processing logic, that is, first obtain the partition and then filter the partition, so maybe we can use one interface for both to control partition related operations. What do u think :)

Reply:
In addition, Danny, what's your opinion for the logic of incremental partition acquisition? 
Option1 : Record a metadata field in the commit to indicate where the last processing was done
Option2: Directly obtain the last completed table service commit time as the new starting point

Option2 is simpler and does not require modifying and processing commit metadata fields.